### PR TITLE
Additional headers for WS `accept` message.

### PR DESCRIPTION
**Pull request by matiuszka**
Hello,

Here is a proposal for improvement about supporting additional headers for the WebSocket `accept` messages.
As per disscuision in gitter:

> Hello,
I recently faced the problem with sending custom headers in WS accept messages. According to ASGI, it should be possible:
https://asgi.readthedocs.io/en/latest/specs/www.html#accept-send-event
I've created an issue and MR for Hypercorn to support that:
pgjones/hypercorn!66
although I am forced to use bare send for this.
Do you think that it can be supported by Starlette's API? If yes, let me know I can create PR.
Here is the proposal: matiuszka/starlette@a080bf5

### CODE REVIEW

**Comment by Kludex** [comment]
It would be cool to have this supported by unicorn as well (although is not a prerequisite).

**Comment by matiuszka** [comment]
@Kludex I will handle it soon.

**Comment by matiuszka** [comment]
Here is PR for `uvicorn`: https://github.com/encode/uvicorn/pull/1293

File: tests/test_websockets.py [hunk 1]
```diff
@@ -301,6 +301,20 @@ async def asgi(receive, send):
...
+        async def asgi(receive, send):
+            websocket = WebSocket(scope, receive=receive, send=send)
+            await websocket.accept(headers=[(b"additional", b"header")])
+            await websocket.close()
+
+        return asgi
+
+    client = test_client_factory(app)
+    with client.websocket_connect("/") as websocket:
+        websocket.additional_headers = [(b"additional", b"header")]
```
Comment (Kludex) [inline tests/test_websockets.py]:
```suggestion
        assert websocket.additional_headers == [(b"additional", b"header")]
```

**Review by Kludex — CHANGES_REQUESTED** [review]
Just a single comment. Thanks @matiuszka ! :smile:

File: starlette/testclient.py [hunk 2]
```diff
@@ -298,6 +298,7 @@ def __init__(
         self.app = app
         self.scope = scope
         self.accepted_subprotocol = None
+        self.additional_headers = None
```
Comment (Kludex) [inline starlette/testclient.py]: Can we use `extra_headers` instead of `addional_headers` to match uvicorn?

File: docs/websockets.md [hunk 3]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (Kludex) [inline docs/websockets.md]: Would you mind readding this line?

**Review by Kludex — CHANGES_REQUESTED** [review]
Besides that, everything is cool!

Thanks @matiuszka ! :smile:

File: tests/test_websockets.py [hunk 4]
```diff
@@ -301,6 +301,20 @@ async def asgi(receive, send):
...
+        async def asgi(receive, send):
+            websocket = WebSocket(scope, receive=receive, send=send)
+            await websocket.accept(headers=[(b"additional", b"header")])
+            await websocket.close()
+
+        return asgi
+
+    client = test_client_factory(app)
+    with client.websocket_connect("/") as websocket:
+        assert websocket.additional_headers == [(b"additional", b"header")]
```
Comment (Kludex) [inline tests/test_websockets.py]:
```suggestion
        assert websocket.extra_headers == [(b"additional", b"header")]
```

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
Thanks @matiuszka ! :smile:

File: starlette/testclient.py [hunk 5]
```diff
@@ -315,6 +316,7 @@ def __enter__(self) -> "WebSocketTestSession":
             self.exit_stack.close()
             raise
         self.accepted_subprotocol = message.get("subprotocol", None)
+        self.extra_headers = message.get("headers", None)
```
Comment (Kludex) [inline starlette/testclient.py:319]: This should probably be `message.get("headers", [])` with #1422

Comment (aminalaee) [reply starlette/testclient.py:319]: I think this part is ok. When passing `None` you'll need to check `assert websocket.extra_headers is None` but with that change:

```python
def test_additional_headers(test_client_factory):
     def app(scope):
         async def asgi(receive, send):
             websocket = WebSocket(scope, receive=receive, send=send)
             await websocket.accept(headers=None)
             await websocket.close()

         return asgi

     client = test_client_factory(app)
     with client.websocket_connect("/") as websocket:
         assert websocket.extra_headers == []
```

Am I missing something?

Comment (Kludex) [reply starlette/testclient.py:319]: You're not. I don't have strong feelings on this. :+1:

**Pull request merged**